day2/src/main.rs

```diff
@@ -0,0 +1,59 @@
+use itertools::Itertools;
+use std::time::Instant;
+
+const INPUT: &[u8] = include_bytes!("input.txt");
+
+fn parse(array: &[u8]) -> i32 {
+    if array.len() == 1 {
+        return (array[0] - b'0') as i32;
+    }
+    return (array[0] - b'0') as i32 * 10 + (array[1] - b'0') as i32;
+}
+
+fn report_is_safe(report: &Vec<i32>) -> bool {
+    let asc = report[1] > report[0];
+    report
+        .windows(2)
+        .find(|w| {
+            let diff = if asc { w[1] - w[0] } else { w[0] - w[1] };
+            diff < 1 || diff > 3
+        })
+        .is_none()
+}
+
+fn part1() -> usize {
+    INPUT
+        .split(|&b| b == b'\n')
+        .map(|report| report.split(|&b| b == b' ').map(parse).collect())
+        .filter(report_is_safe)
+        .count()
+}
+
+fn part2() -> i32 {
+    let mut num_safe = 0;
```

Comment:
Minor thing as it might be less legible but you could possibly do something similar to part 1 with a `filter` and `count` rather than having this mutable variable.

Primarily a stylist choice but there are some cases where Rust is able to perform some optimisations across iterator chains that it can't detect in `for` loops, but I couldn't say if that applies here without benchmarking.

Reply:
I tried this at first but was getting a type mismatch. The final `any` call had this error
```
type mismatch in closure arguments
expected closure signature `fn(Vec<&_>) -> _`
   found closure signature `fn(Vec<_>) -> _`
```
so I guess `combinations` was returning a Vec of references rather than a Vec of i32, so they'd have to be cloned?

Reply:
I'd have to see the full code to really grok what was going on but at a guess you might have been using an `iter` rather than `into_iter` and collecting over that? Since you're working with `i32`s here it would be very cheap to called `.cloned()` in the interator chain before `combinations` was called - which would remove the reference, there might be a neater way but I'd need to see the full code to be sure.

Anyway not a biggie.